### DISCUSSION TITLE
Low Bit Rate Devices

### DISCUSSION THREAD

**Discussion by OtisByron** [markqvist/Reticulum#11] [Ideas]
This discussion is not intended to be amateur radio specific.

The manual indicates that the minimum bit rate for Reticulum is around 1000 b/s. That is quite low by normal standards, but it would be nice to go much lower.

I'm thinking specifically of interfacing with HF radios to operate as a backbone network to facilitate messaging, using something like NomadNet. Using a HF radio and a NVIS antenna configuration you could extend your network coverage throughout a region several hundred miles in diameter.

Many modems typically used for long range HF communication are operating in the range of 10's of bits per second (BPSK-31, Contestia 4/250, etc). Even HF packet is 300 baud, such as with the OpenModem.

So the questions are:

What are the challenges Reticulum faces with such a low data rate?

Would it be feasible to facilitate extremely low data rates, such as 31 b/s?

As a side note, I'm envisioning the use of soft modems such as Fldigi or minimodem in this scenario. I have played around with "side loading" a custom Reticulum interface that interfaces with Fldigi (using pyfldigi) and uses a HDLC-ish approach to parse data out of the background noise characters, and it does work conceptual.

**Comment by samuk** [comment]
Interesting. Would http://js8call.com/ be a good candidate for an HF carrier?

**Reply by OtisByron** [reply]
@samuk

I quite enjoy JS8Call, and it is my current go-to for amateur radio when I have a computer handy, but I think there may be a few issues that might make this an overly complex option to use with Reticulum:

1. The necessity to have accurate time (within 2 seconds, per the faq on the website). I've read a little about using the time offset tracked by JS8Call itself to adjust your local computers time (which is genius and quite interesting), but I think time keeping in general would be counter-productive to the Reticulum end goal of running on low power embedded devices.

2. The fixed transmission window may be a challenge. This would also effectively limit your bit rate even further.

3. The API is poorly documented (I would love to be corrected on that statement!). If the API was documented that may or may not chnage #2.

4. The networking features of JS8Call would be redundant to Reticulum, and would add overhead and complexity.

JS8Call is the best solution I am aware of today for HF messaging in a standalone setup. However, I think there would be value in joining HF and local networks over Reticulum.

This is the scenario I am imagining:
- Two (or more) remote nodes connected over a HF data link
- Each HF node is also connected to a RNode device
- Communities hundreds or thousands of miles away from each other could be connected so that each end node could send a message to any other local or remote node

**Reply by markqvist** [reply]
> Two (or more) remote nodes connected over a HF data link. Each HF node is also connected to a RNode device. Communities hundreds or thousands of miles away from each other could be connected so that each end node could send a message to any other local or remote node

This is exactly the kinds of scenarios I have been seeking to make possible with Reticulum. Currently, the only realistic way to reliably interconnect communities separated by large distances is via HF. With the current stack of Reticulum -> LXMF -> NomadNet, this is possible, in theory. The two systems running both an RNode and a HF link (of some sort) could each run a LXMF Propagation Node (which is activated automatically in NomadNet as soon as you enable node hosting). In such a case, the two propagation nodes would sync messages between each other, to and from the two communities when the conditions allows for the link to exist, since LXMF propagation nodes are tolerant to quite a lot of intermittency over their peering links.

As you have correctly pointed out @OtisByron, the only missing part of that puzzle is making Reticulum work reliably over HF.

I am known to be a bit slow in updating some documentation, and the _real_ lower limit for Reticulum right now is actually around 475 bits per second. Pretty close to the 300 baud barrier of HF packet at least, and I am completely confident that we can _at least_ go there.

Luckily, the issue is not so much with the fundamentals of Reticulum itself. I actually designed that to be completely time-invariant, and you could run it on a 1 bit per week link, if you had the time for that. The problem is the _range_ of transfer speeds. Currently, Reticulum works gracefully from about 475 bits per second up to about 100 megabits per second. To go higher is not really difficult, it is mostly a question of throwing more CPU or more optimisation of the crypto primitives at it, but I don't really feel like that is in any way necessary at this point. 100 megabits is _plenty_ fast right now. What we want is to go slower.

For this to happen, the algorithms that predict and adjusts timeouts need to be modified to support this extremely broad range of expected transfer speeds. Latencies increase exponentially in this territory, and I can't just allow Reticulum to wait for the amount of time a request _might_ take on a 35bps link as a standard case. Ultimately, it would mean users sitting on a relatively fast link, having to wait several minutes for timeouts on unreachable destinations, for example, when in reality they could have gotten notified of this in a few seconds.

So the only option is working out a solid theory for dealing with this, admittedly rather extreme, range of intended bitrates, and adjusting the expectations of Reticulums timing algorithms accordingly and correctly. I've been doing quite a lot of headscratching around this challenge for a while now, and while I do not yet have a solid solution, I am confident that one exists, and that it is something we can realise.

But it _is_ a nontrivial problem, since we can't just throw in a bunch of probing and signalling to determine the speeds and latencies. That would waste all the very precious bandwidth before we even got started, so it has to be almost completely passive inference, as with the rest of Reticulum.

**Reply by OtisByron** [reply]
Thanks for the detailed response @markqvist. The fact that this type of situation is on the road map for Reticulum is encouraging. I shouldn't be surprised, because for the past few years whenever I've been thinking about a way to improve the system, that feature comes out in the next release :)

What you mentioned about timing algorithms in Reticulum being the sticking point due to the wide range of data rates makes a lot of sense, as well as the need for a passive approach. I'll poke around in the documentation and the code with that in mind and I'll add any questions or comments to this discussion topic.

On a related but slightly different topic I've been working on a hardware design to integrate all the necessary parts/pieces for a multi-purpose node, like the one discussed in this thread, into a single piece of portable equipment. I don't want to muddy the waters on this repo if you don't think this is the right place to discuss such a topic. Let me know if you have any thoughts or recommendations on the appropriate location for such a discussion.

**Reply by markqvist** [reply]
My pleasure, and thanks for the words of encouragement.

Your work on a portable multi-purpose node sounds very interesting. Generally, I think any discussion around both Reticulum and related projects belong in this forum, so don't hesitate to bring it up here if you want to. It's easier to collect all the information here, instead of spreading it out over the discussion pages of multiple repos I think, as long as it is not directly related to the code and development of those other projects.

**Comment by OtisByron** [comment]
As mentioned by @markqvist this is definitely a non-trivial problem. My comments here are not a solution to the broader issue, but a possible solution to add some limited support for low data rate devices in the short term.

This concept would support data rates down to around 31 bps, but only over links. The basic idea is to leverage packet RTT which already exists for links. Once the RTT is known for a link, packets sent over the link will use the RTT to calculate their timeout. This also means no additional network traffic, since this process already happens during link establishment.

The only holdup I can see at the moment is the link establishment timeout, which is currently set to the default global timeout per hop (5 seconds) multiplied by the number of hops to the destination. If the logic in Link.__watchdog_job() for the case Link.status == Link.HANDSHAKE was adjusted to add 60 seconds to the handshake timeout this should allow the necessary time for a low data rate device to communicate the necessary 237 bytes (per Reticulum documentation) to establish the link and set the link RTT.

31.25 bits per second = 0.256 seconds per byte
0.256 s/byte * 237 bytes = 60.7 seconds

This 60 seconds would be added to the normal establishment timeout per hop to allow additional timeout for other hops along the path. 

In theory, this approach should add enough time for 1 hop over a 31 bps interface, or up to ~10 hops over a 300 bps interfaces. 

300 bits per second = 0.0267 seconds per byte
0.0267 s/byte * 237 bytes = 6.3 seconds per hop

Best I can tell, announce packets would still propagate over a low data rate link, so I don't think node discovery should be an issue.

I'm sure I've missed something, but hopefully the basic idea comes across. This is far from a perfect solution, but it would allow Reticulum to support low data rate devices in some capacity with minimal development effort, no impact to normal higher speed traffic, and no additional network traffic. I believe this would also allow LXMF and NomadNet to work over a low data rate device as well.

Side note: I'm ignoring the case in Link.__watchdog_job() where Link.status == Link.PENDING because it doesn't seem like that case ever occurs. When a link is created, either as the initiator or via Link.validate_request(), Link.handshake() is called before Link.start_watchdog().

**Reply by OtisByron** [reply]
Alternatively, just adding 15 seconds to the link establishment timeout should allow up to two 300 bps packet radio hops along a path.

**Reply by markqvist** [reply]
The basic idea you have is sound, but the problem is that adding that extra time for all link types, and in many cases, adding a whole minute of waiting time before getting notified that link establishment failed is not good, so we really need some way to set this sensibly taking the entire path into consideration. You could however modify the code as you suggested locally, and experiment with how well it works. I would be very interested in hearing your results.

I think I have a workable solution to the problem though, but it requires a bit more thinking through. It involves making the concept of "hops" a bit more abstract, so that very low bandwidth links can be represented as higher hop values. I'm currently experimenting with the feasibility of this.

**Reply by ghost** [reply]
The system you describe reminds me of the [Delta-t protocol](https://www.osti.gov/servlets/purl/5542785-VVKXmp/), a robust timer-based mechanism for link reliability and sequencing that simply uses 3 timers for each link calculated with measured RTT.

Most people have forgotten about this protocol -- it was made before I was born -- so there aren't any implementations around today. But its concepts are simple, even at times enlightening.

**Reply by markqvist** [reply]
Now that was interesting! I did not know about that. I will give it a look, probably a lot to learn there!

**Comment by markqvist** [comment]
Everyone in this thread might find it interesting that the commits I pushed yesterday and today lays down the basic support for timing independence, and will soon allow *absurdly* low bitrates, if you have the patience for it ;) We are pretty close to HF compatibility now.

**Comment by OtisByron** [comment]
@markqvist would you be able to comment on the minimum bit rate supported by Reticulum at this time?

The more I have thought about Reticulum over HF the more I agree with @samuk that JS8Call is the best solution currently available, even though it comes with its challenges. The extremely low bit rate would likely limit real world usage to simple messaging, but the performance mitigates a lot of the challenges that come with HF which makes it more functional for the average user.

**Reply by markqvist** [reply]
It is still at 475bps. But I will probably add a per-interface option in one of the next releases, that allows you to specify an extra timeout delay for just those. That should make it work, at least provisionally, for comms where the first hop is <475bps.
